contracts/hooks/FullRange.sol

```diff
@@ -0,0 +1,430 @@
+                // position.tokensOwed1 -= delta.amount1();
+            }
+
+            if (data.key.currency1.isNative()) {
+                poolManager.settle{value: uint128(-delta.amount1())}(data.key.currency1);
+            } else {
+                poolManager.settle(data.key.currency1);
+            }
+        }
+
+        return abi.encode(delta);
+    }
+
+    function getHooksCalls() public pure override returns (Hooks.Calls memory) {
+        return Hooks.Calls({
+            beforeInitialize: true,
+            afterInitialize: false,
+            beforeModifyPosition: true,
+            afterModifyPosition: false,
+            beforeSwap: true,
+            afterSwap: false,
+            beforeDonate: false,
+            afterDonate: false
+        });
+    }
+
+    function addLiquidity(
+        address tokenA,
+        address tokenB,
+        uint24 fee,
+        uint256 amountADesired,
+        uint256 amountBDesired,
+        address to,
+        uint256 deadline
+    ) external ensure(deadline) returns (uint128 liquidity) {
+        IPoolManager.PoolKey memory key = IPoolManager.PoolKey({
+            currency0: Currency.wrap(tokenA),
+            currency1: Currency.wrap(tokenB),
+            fee: fee,
+            tickSpacing: 60,
+            hooks: IHooks(address(this))
+        });
+
+        // replacement addLiquidity function from LiquidityManagement.sol
+        (uint160 sqrtPriceX96,,,,,) = poolManager.getSlot0(key.toId());
+
+        if (sqrtPriceX96 == 0) revert PoolNotInitialized();
+
+        liquidity = LiquidityAmounts.getLiquidityForAmounts(
+            sqrtPriceX96,
+            TickMath.getSqrtRatioAtTick(MIN_TICK),
+            TickMath.getSqrtRatioAtTick(MAX_TICK),
+            amountADesired,
+            amountBDesired
+        );
+
+        modifyPosition(
+            key,
+            IPoolManager.ModifyPositionParams({
+                tickLower: MIN_TICK,
+                tickUpper: MAX_TICK,
+                liquidityDelta: int256(int128(liquidity))
+            })
+        );
+
+        // NOTE: we've already done the rebalance here
+
+        Position.Info memory posInfo = poolManager.getPosition(key.toId(), address(this), MIN_TICK, MAX_TICK);
+
+        PoolInfo storage poolInfo = poolToInfo[key.toId()];
+
+        poolInfo.tokensOwed0 += uint128(
+            FullMath.mulDiv(
+                posInfo.feeGrowthInside0LastX128 - poolInfo.feeGrowthInside0LastX128,
+                poolInfo.liquidity,
+                FixedPoint128.Q128
+            )
+        );
+        poolInfo.tokensOwed1 += uint128(
+            FullMath.mulDiv(
+                posInfo.feeGrowthInside1LastX128 - poolInfo.feeGrowthInside1LastX128,
+                poolInfo.liquidity,
+                FixedPoint128.Q128
+            )
+        );
+
+        poolInfo.feeGrowthInside0LastX128 = posInfo.feeGrowthInside0LastX128;
```

Comment:
maybe we discussed,  but, why are we recording this? In v3-periphery we need to keep track for each individual, but if this sits in the pool since it's at the pool level, I think it's just redundant to keep it here.